### PR TITLE
Support inner types and enums with custom constructors for bytecode generator

### PR DESCRIPTION
**Pull request by dstepanov**

### CODE REVIEW

File: test-suite-bytecode/src/test/java/io/micronaut/sourcegen/example/InnerTypesTest.java [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+
+//        MyEnumWithInnerTypes.InnerRecord innerRecord = MyEnumWithInnerTypes$InnerRecordBuilder.builder().id(3).build();
+//        assertEquals(3, innerRecord.id());
+
+        MyEnumWithInnerTypes.InnerClass innerClass = new MyEnumWithInnerTypes.InnerClass("name");
+        assertEquals("name", innerClass.getName());
+    }
+
+//    @Test
+//    public void recordTest() {
```
Comment (dstepanov) [inline test-suite-bytecode/src/test/java/io/micronaut/sourcegen/example/InnerTypesTest.java:46]: Records not supported yet, keeping the commented out tests for the future

File: sourcegen-model/src/main/java/io/micronaut/sourcegen/model/ClassTypeDef.java [hunk 2]
```diff
@@ -54,14 +54,24 @@ public sealed interface ClassTypeDef extends TypeDef {
      * @return The simple name
      */
     default String getSimpleName() {
-        return NameUtils.getSimpleName(getName());
+        String name = getCanonicalName();
+        int i = name.lastIndexOf('.');
+        if (i == -1) {
+            return name;
+        }
+        return name.substring(i + 1);
```
Comment (elifKurtay) [inline sourcegen-model/src/main/java/io/micronaut/sourcegen/model/ClassTypeDef.java:62]: return name.substring(name.lastIndexOf('.') + 1);

**Comment by elifKurtay** [comment]
Right now when we annotate any TypeDef, we cannot reach the Type's information inside AnnotatedClassTypeDef/AnnotatedTypeDef. It might be nice to add getters to reach the typeDef inside or copy the ClassTypeDef's default functions also here like in Parameterized.

**Review by elifKurtay — APPROVED** [review]

**Review by graemerocher — APPROVED** [review]

**Comment by dstepanov** [comment]
> Right now when we annotate any TypeDef, we cannot reach the Type's information inside AnnotatedClassTypeDef/AnnotatedTypeDef. It might be nice to add getters to reach the typeDef inside or copy the ClassTypeDef's default functions also here like in Parameterized.

I'm not sure what you mean. In most cases, I only differentiate whether the type is `TypeDef` or `ClassTypeDef`; that's why we have two annotated variations. What is precisely the use case?

**Pull request merged**